File: embabel-database-agent/src/main/java/com/embabel/database/agent/service/ModelSuggestionService.java
/*
 * Copyright 2024-2025 Embabel Software, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.embabel.database.agent.service;

import java.util.*;

import com.embabel.agent.api.common.autonomy.AgentInvocation;
import com.embabel.agent.core.*;
import com.embabel.database.agent.domain.*;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.embabel.agent.domain.io.UserInput;

/**
 * service that coordinates agents to make model suggestions
 */
@Service
public class ModelSuggestionService {
    
    private static final Log logger = LogFactory.getLog(ModelSuggestionService.class);

    @Autowired
    AgentPlatform agentPlatform;

    @Autowired
    SessionManagementService sessionManagementService;

    public SessionContext getProviderSuggestions(String request) throws Exception {
        //return the session ID and trigger the process
        //establish a session
        String sessionId = UUID.randomUUID().toString();        
        //build the invocation
        AgentInvocation<Providers> agentInvocation = AgentInvocation.builder(agentPlatform).build(Providers.class);
        //run the process
        AgentProcess agentProcess = agentInvocation.run(Collections.singletonMap("userInput",new UserInput(request)));
        //wait until it's finished
        while (!agentProcess.getFinished()
                && !agentProcess.getStatus().name().equalsIgnoreCase("STUCK")) {
            logger.info(agentProcess.getStatus().name());
            Thread.sleep(500);//TODO externalize
        } //end while
        if (agentProcess.getStatus().name().equalsIgnoreCase("STUCK")) {
            //prematurely stopped
            logger.warn("stuck");
            return null;
        } //end if
        //store the process for later blackboard retrieval
        Blackboard blackboard = agentProcess.getProcessContext()
                .getBlackboard();
        //save against the session
        Object result = agentProcess.lastResult();
        if (result == null) {
            logger.info("No suggestions");
            return null;
        } //end if
        Providers providers = (Providers) result;
        SessionContext sessionContext = new SessionContext(sessionId,providers,blackboard,request);
        //add to the session management
        sessionManagementService.save(sessionContext);
        //return the results
        return sessionContext;
    }

    private static final String USERINPUT = "userInput";
    private static final String PREVIOUS_PROMPT = "previousPrompt";

    public ModelSuggestion getModelSuggestion(String provider,String sessionId) throws Exception {
        //get the previous session
        SessionContext sessionContext = sessionManagementService.findById(sessionId);
        //check
        if (sessionContext == null) {
            throw new IllegalArgumentException("session Id is invalid " + sessionId);
        } //end if
        //retrieve the blackboard
        Blackboard blackboard = sessionContext.blackboard();
        //build the process options
        ProcessOptions processOptions = ProcessOptions.builder().
                blackboard(blackboard)
                .build();
        //build the invocation
        AgentInvocation<ModelSuggestion> modelSuggestionAgentInvocation = AgentInvocation.builder(agentPlatform)
                .options(processOptions)
                .build(ModelSuggestion.class);
        //TODO clean up process for sessions
        //get the suggestion
        return modelSuggestionAgentInvocation.invoke(Map.of(USERINPUT,new UserInput(provider),PREVIOUS_PROMPT,sessionContext.prompt()));
    }
}

File: embabel-database-server/src/main/kotlin/com/embabel/database/server/service/ModelSuggestionService.kt
/*
 * Copyright 2024-2025 Embabel Software, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.embabel.database.server.service

import com.embabel.agent.core.AgentPlatform;
import com.embabel.agent.core.AgentProcess
import com.embabel.agent.domain.io.UserInput;

import com.embabel.common.ai.model.ModelMetadata

import com.embabel.database.agent.domain.ProviderList
import com.embabel.database.agent.domain.TagList
import com.embabel.database.core.repository.AiModelRepository
import com.embabel.database.core.repository.LlmModelMetadata

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import org.slf4j.LoggerFactory

import java.util.concurrent.ConcurrentHashMap

public object RequestRegistry {

    val processMap: MutableMap<String, Map<String,Any>> = ConcurrentHashMap()
}

@Service
class ModelSuggestionService {

    private val logger = LoggerFactory.getLogger(ModelSuggestionService::class.java)

    val agentName = "ModelSuggestionAgent"

    //reuse the existing agent execution service
    @Autowired
    lateinit var agentExecutionService: AgentExecutionService

    @Autowired
    lateinit var aiModelRepository: AiModelRepository

    fun getProviders(prompt: String, requestId: String) : String? {
        //execute the agent
        //wait
        //get response objects
        //posit back to the user
        //build the input
        val userInput = UserInput(prompt)
        val options = mapOf("userInput" to userInput)
        var agentProcess = agentExecutionService.createProcess(agentName,options)
        //run
        agentExecutionService.runAgentProcessAsync(agentProcess)
        //monitor
        agentExecutionService.monitor(agentProcess)
        //register this result
        var executionMap = mapOf("prompt" to prompt, "process" to agentProcess)
        RequestRegistry.processMap[requestId] = executionMap
        //complete - get status and objects
        var results = agentProcess.lastResult()
        var providerList = results as? ProviderList
        // return
        return providerList?.providers?.joinToString(",")
    }

    fun getModels(provider: String, requestId: String) : List<ModelMetadata>? {
        //use the provider (string) and the request id to retrieve the tags
        var process = RequestRegistry.processMap[requestId]
        //search for models from the repository for the provider
        var models: List<ModelMetadata>? = aiModelRepository.findByProvider(provider)
        //now validate for the tag
        var processMap = process as? Map<String,Any>
        var history = processMap?.get("process") as AgentProcess
        var objectList = history?.objects
        //loop
        val tagList: TagList? = objectList?.filterIsInstance<TagList>()?.firstOrNull()
        val tagsArray: Array<String>? = tagList?.tags?.toTypedArray()
        //filter
        val results = models?.filterIsInstance<LlmModelMetadata>()?.filter { model -> model.tags?.any { it in (tagsArray ?: emptyArray())} == true }.orEmpty()
        //return
        return results.takeIf { it.isNotEmpty() }
    }
}
